### ISSUE TITLE
[trainer] RLOO scalar and vectorized advantage disagree on singleton groups

### ISSUE THREAD

**Issue by rrrxxx0510** [verl-project/verl#7793]
The scalar and vectorized RLOO advantage implementations disagree on singleton groups (groups with a single sample), which can arise when filtering (e.g. overlong or dynamic filtering) drops all but one sample of a prompt group.

- `compute_rloo_outcome_advantage` (scalar): a singleton group keeps the raw score as the advantage (`id2mean = 0`, the update loop is skipped), so the surviving sample still gets a learning signal. This matches the GRPO singleton convention in `compute_grpo_outcome_advantage` (mean=0, std=1).
- `compute_rloo_vectorized_outcome_advantage` (vectorized): the final `* (c > 1)` zeroes the advantage, so the surviving sample contributes no gradient at all.

Both estimators are selected via `algorithm.adv_estimator` and are otherwise mathematically equivalent (I verified the leave-one-out algebra for multi-sample groups), so silently differing behavior on singletons looks unintended. Example with `n_samples_per_prompt=2` where one sample gets filtered:

```
group scores: [0.7]            # singleton after filtering
scalar:      advantage = 0.7   # trains on the surviving sample
vectorized:  advantage = 0.0   # sample contributes nothing
```

Suggested fix: pass the raw score through for singleton groups in the vectorized path (e.g. `adv = torch.where(c > 1, adv, scores)`), aligning it with the scalar implementation and the GRPO singleton convention. Happy to send a PR.

**Comment by linhongyu510** [comment]
Adjacent finding while looking at this, kept deliberately separate: the same function also loses the reward dtype, independently of the singleton behaviour discussed here.

`torch.bincount(weights=...)` accumulates in float64 whatever the weights' dtype is, so the group sums promote the whole expression. With **every group holding two samples** (no singletons involved):

```
index = ["p0","p0","p1","p1"]

input torch.float32    scalar->torch.float32    vectorized->torch.float32
input torch.bfloat16   scalar->torch.bfloat16   vectorized->torch.float64
input torch.float16    scalar->torch.float16    vectorized->torch.float64
```

So `rloo` and `rloo_vectorized` disagree on output dtype even where they agree numerically. `test_rloo_and_vectorized_equivalence` can't see it because it builds rewards as `torch.randn(..., dtype=torch.float32)`, and fp32 is the one dtype that's unaffected.

Sent as #7964 — a one-line `.to(scores.dtype)` plus a dtype test that uses no singleton groups, so it doesn't depend on how this issue gets resolved. I checked that it composes with the `torch.where(c > 1, loo, scores)` fix from #7581 / #7794 rather than competing with it (applied both locally, 26 passed); the lines are adjacent so there may be a trivial textual conflict depending on merge order, and I'm happy to rebase on whichever lands first.

On the issue itself: I'd arrived at the same conclusion and the same `torch.where` before finding this thread, and I think your diagnosis is right — the scalar path's `id2mean = 0` and `group_mean_std`'s explicit `mean=0, std=1` singleton branch are the same convention, so the vectorized RLOO is the odd one out. Worth adding that it isn't only a filtering edge case: `rollout.n` defaults to `1`, which makes *every* group a singleton, so on the default config the vectorized estimator zeroes the entire batch. No point in a fourth PR for it though.
